Drivers/Src/MotorController.c

```diff
@@ -1,124 +1,214 @@
  * @return  SUCCESS if a message is read
  */ 
-ErrorStatus MotorController_Read(CANbuff *message, car_state_t *car){
+ErrorStatus MotorController_Read(CANbuff *message){
 
     uint32_t id;
     uint8_t data[8] = {0};
-    uint32_t length = BSP_CAN_Read(CAN_3, &id, data);
     uint32_t firstSum = 0;
     uint32_t secondSum = 0;
+    CPU_TS ts;
+	OS_ERR err;
+	
+	// Check to see if a mailbox is available: BLOCKING
+	OSSemPend(
+        &MotorController_ReceiveSem4,
+	    0,    
+		OS_OPT_PEND_BLOCKING,
+		&ts,
+		&err);
+	assertOSError(0, err);
+    uint32_t length = BSP_CAN_Read(CAN_3, &id, data);
 
     if(length>0){
         message->id = id;
         //get first number (bits 0-31)
         for(int j = 0; j < MAX_CAN_LEN/2; j++){
+            firstSum <<= 8;
             firstSum += data[j];
-            if(j != MAX_CAN_LEN/2 - 1){
-                firstSum = firstSum << 8;
-            }
         }
         //get second number (bits 32-63)
         for(int k = MAX_CAN_LEN/2; k < MAX_CAN_LEN; k++){
+            secondSum <<= 8;
             secondSum += data[k];
-            if(k != MAX_CAN_LEN-1){
-                secondSum = secondSum << 8;
-            }
         }
         message->firstNum = firstSum;
         message->secondNum = secondSum;
-
-        union {
-            uint32_t n;
-            float f;
-        } convert;
         
         switch (id) {
             // If we're reading the output from the Motor Status command (0x241) then 
             // Check the status bits we care about and set flags accordingly
             case MOTOR_STATUS: {
-                if(MASK_MOTOR_TEMP_ERR & firstSum)
-                {
-                    car->MotorErrorCode.motorTempErr = ON;
+                if(MASK_MOTOR_TEMP_ERR & firstSum){
+                    Motor_FaultBitmap |= T_TEMP_ERR;
                 }
 
-                if(MASK_SS_ERR & firstSum)
-                {
-                    car->MotorErrorCode.slipSpeedErr = ON;
+                if(MASK_SS_ERR & firstSum){
+                    Motor_FaultBitmap |= T_CC_VEL_ERR;
                 }
 
-                if(MASK_CC_ERR & firstSum)
-                {
-                    car->MotorErrorCode.CCVelocityErr = ON;
+                if(MASK_CC_ERR & firstSum){
+                    Motor_FaultBitmap |= T_SLIP_SPEED_ERR;
                 }
 
-                if(MASK_OVER_SPEED_ERR & firstSum)
-                {
-                    car->MotorErrorCode.overSpeedErr = ON;
+                if(MASK_OVER_SPEED_ERR & firstSum){
+                    Motor_FaultBitmap |= T_OVER_SPEED_ERR;
                 }
+
+                assertTritiumError(Motor_FaultBitmap);
+                
                 break;
             }
             case MOTOR_DRIVE: {
-                convert.n = secondSum;
-                car->CurrentVelocity = convert.f;
+                CurrentVelocity = *(float *) &secondSum;
                 break;
```

Comment:
I don't think we'll ever recieve a MOTOR_DRIVE command, we only send it. I don't see handling for the actual messages we receive as outlined in the tritium datasheet.

Reply:
We will be receiving a MOTOR_VELOCITY message, I think the case statement got messed up here.